{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transfer gene hits from one bed file to another\n",
    "\n",
    "### This notebook reads blastn/blastp results and transfers genes from the reference bed file to the needed bed file. To get these results, follow Viera's instructions:\n",
    "\n",
    ">1.) B. spiz. - prepare CDS (Coding DNA Sequence) fasta file using one of the commands below (Sometimes these files are attached in the NCBI reference annotation DB and can be downloaded):\n",
    ">\n",
    ">a.) bedtools getfasta\n",
    "bedtools getfasta -fi reference.fasta -bed W23wt_June2021.bed -fo cds_sequences.fasta\n",
    ">\n",
    ">OR\n",
    ">b.) Convert BED to GFF Format, use 'extractfeat' function (tool) from EMBOSS package extractfeat -sequence reference.fasta -feat CDS -source annotation.gff -outseq cds_sequences.fasta\n",
    ">\n",
    ">2.)  Compare CDS from B. subtilis and B. spiz. using NCBI blastn (e-value should be 1e-6 at least; call only the best hit)\n",
    ">\n",
    ">-> Instead of the whole genome reference_genome.fasta for B. subtilis,\n",
    ">use also here CDS file for both species. The headers of B. subtilis CDS sequences should contain IDs of B. subtilis genes.\n",
    ">\n",
    ">a.) build reference for B. subtilis\n",
    ">\n",
    ">makeblastdb -in reference_genome_CDS.fasta -dbtype nucl -out reference_db_CDS\n",
    ">\n",
    ">b.) compare (query) CDS of B. spiz. to the reference genome; one best hit per B. spiz. CDS sequence\n",
    ">\n",
    ">blastn -query cds_sequences.fasta -db reference_db_CDS -out blast_results.txt -outfmt 6 -evalue 1e-6 -max_target_seqs 1\n",
    ">\n",
    ">Sequences not found in the blastn results are very specific to B.  \n",
    ">spiz. Use B. subt. IDs from the results and add them to the B. spiz.  \n",
    ">annotation file (bed, gtf).\n",
    ">\n",
    ">Do you have a conversion table of gene IDs used for B. subtilis (NCBI, Ensembl, other common)?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Executed\n",
    "```bedtools getfasta -fi W23wt.fasta -bed W23wt_June2021.bed.txt -fo W23wt_cds_sequences.fasta\n",
    "bedtools getfasta -fi Bs166NCe.fasta -bed Bs166NCe_June2021.bed.txt -fo Bs166NCe_cds_sequences.fasta\n",
    "bedtools getfasta -fi \"/mnt/c/Users/Philipp/Downloads/GCF_000009045.1/GCF_000009045.1_ASM904v1_genomic.fna\" -bed Bs166NCe_June2021.bed.txt -fo Bs166NCe_cds_sequences.fasta\n",
    "bedtools getfasta -fi \"/mnt/c/Users/Philipp/Downloads/GCF_006094475.1/GCF_006094475.1_ASM609447v1_genomic.fna\" -bed W23wt_June2021.bed.txt -fo W23wt_cds_sequences.fasta\n",
    "\n",
    "makeblastdb -in Bsub_GCF_000009045.1_ASM904v1_genomic.fna -dbtype nucl -out Bsub_reference_db_CDS\n",
    "\n",
    "blastn -query Bsub_GCF_000009045.1_ASM904v1_genomic.fna -db Bsub_reference_db_CDS -out blast_results.txt -outfmt 6 -evalue 1e-6 -max_target_seqs 1\n",
    "\n",
    "\n",
    "\n",
    "# to get blastp database\n",
    "makeblastdb -in Bsub_GCF_000009045.1_ASM904v1_genomic.fna -dbtype prot -out Bsub_protein_reference_db_CDS\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual code\n",
    "\n",
    "## Ansatz\n",
    "\n",
    "- iterate through the bed file you want to annotate\n",
    "- for each line get the start and stop position\n",
    "- check if the interval is found in blast results (+/- 3 nucleotides)\n",
    "    - 3rd column = pident = should be > 95?%\n",
    "    - 7th/8th column = start/stop query, position in CDS = useless!!!\n",
    "    - 9th/10th column = start/stop subject, position in CDS = useless!!!\n",
    "    - instead: use qseqid (refering to bed file; iterated line) and sseqid (refering to reference bed file)\n",
    "        - example: chr1:1938-3074\tchr2:1939-3075\n",
    "        - use RegEx magic to extract 1938, 3074 and 1939, 3075\n",
    "- if good hit in blast results\n",
    "    - calculate position on reference genome bed file\n",
    "        - `offset = start subject - start query`\n",
    "    - lookup locus in reference bed file\n",
    "    - copy gene identifier to improved bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the bed file and blast results\n",
    "#incomplete_bed_file = pd.read_csv(\"//wsl.localhost/Ubuntu/home/philipp/dicts/W23wt_June2021.bed.txt\", sep='\\t', header=None, names=['seqName', 'start', 'end', 'direction', 'type1', 'type2', 'name', 'locus_tag', 'product'], comment='#')\n",
    "incomplete_bed_file = pd.read_csv(\"//wsl.localhost/Ubuntu/home/philipp/dicts/Features from W23_NCBI_comK_inducible_delta_comK.bed\", sep='\\t', header=None, names=['seqName', 'start', 'end', 'direction', 'type1', 'type2', 'name', 'locus_tag', 'product'], comment='#')\n",
    "\n",
    "reference_bed_file = pd.read_csv(\"//wsl.localhost/Ubuntu/home/philipp/dicts/Bs166NCe_June2021.bed.txt\", sep='\\t', header=None, names=['seqName', 'start', 'end', 'direction', 'type1', 'type2', 'name', 'locus_tag', 'product'], comment='#')\n",
    "\n",
    "blast_filename = \"//wsl.localhost/Ubuntu/home/philipp/dicts/blast_results_MelihFasta.txt\"\n",
    "#blast_filename = \"//wsl.localhost/Ubuntu/home/philipp/dicts/blast_chr_results.txt\"\n",
    "blast_results = pd.read_csv(blast_filename, sep='\\t', header=None, names=['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore'], comment='#')\n",
    "\n",
    "# Filter blast results for high identity matches\n",
    "blast_results = blast_results[blast_results['pident'] > 80]\n",
    "\n",
    "# create new columns with the real loci\n",
    "blast_results[\"real_qstart\"] = blast_results[\"qseqid\"].str.extract(r'[\\w.-]+:(\\d+)-\\d+').astype(int)\n",
    "blast_results[\"real_qend\"] = blast_results[\"qseqid\"].str.extract(r'[\\w.-]+:\\d+-(\\d+)').astype(int)\n",
    "blast_results[\"real_sstart\"] = blast_results[\"sseqid\"].str.extract(r'[\\w.-]+:(\\d+)-\\d+').astype(int)\n",
    "blast_results[\"real_send\"] = blast_results[\"sseqid\"].str.extract(r'[\\w.-]+:\\d+-(\\d+)').astype(int)\n",
    "\n",
    "#incomplete_bed_file = incomplete_bed_file[0:8]\n",
    "\n",
    "# Create a new DataFrame for the improved bed file\n",
    "##improved_bed = pd.DataFrame(columns=['seqName', 'start', 'end', 'direction', 'type1', 'type2', 'name', 'locus_tag', 'product'])\n",
    "improved_bed = pd.DataFrame(columns=incomplete_bed_file.columns) # copy columns from bed_file\n",
    "improved_bed = improved_bed.astype(incomplete_bed_file.dtypes.to_dict()) # copy dtypes from bed_file\n",
    "\n",
    "# Iterate through the bed file\n",
    "for index, row in incomplete_bed_file.iterrows():\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "    \n",
    "    # Check if the interval is found in blast results (+/- 3 nucleotides)\n",
    "    match = blast_results[(blast_results['real_qstart'] <= start + 3) & (blast_results['real_qend'] >= end - 3)]\n",
    "    #print(\"match: \", start , end,  \"\\n\", match[[\"real_qstart\", \"real_qend\"]])\n",
    "\n",
    "    \n",
    "    if not match.empty:    \n",
    "        #if len(match) > 1:\n",
    "        #    print(f\"Multiple matches found for {row['name']} between {start} and {end}:\\n{match}\")\n",
    "        \n",
    "        # Get the shortest match\n",
    "        blast_row = match.loc[match['length'].idxmin()]\n",
    "        \n",
    "        # Calculate position on reference genome\n",
    "        offset = blast_row['real_sstart'] - blast_row['real_qstart']\n",
    "        ref_start = start + offset\n",
    "        ref_end = end + offset\n",
    "        \n",
    "        # Lookup locus in reference bed file (+/- 3 nucleotides)\n",
    "        ref_locus = reference_bed_file[(reference_bed_file['start'] >= ref_start - 3) & (reference_bed_file['end'] <= ref_end + 3)]\n",
    "        \n",
    "        #if ref_locus.empty:\n",
    "        #    print(print(f\"Found NO locus for {row['name']}\"))\n",
    "        \n",
    "        if not ref_locus.empty:\n",
    "            #print(f\"Found locus for {row['name']}\")\n",
    "            #print(row['seqName'])\n",
    "            # Copy gene identifier to improved bed file\n",
    "            improved_bed = pd.concat([improved_bed, pd.DataFrame({\n",
    "                'seqName': ref_locus.iloc[0]['seqName'],\n",
    "                'start': row['start'],\n",
    "                'end': row['end'],\n",
    "                'direction': ref_locus.iloc[0]['direction'],\n",
    "                'type1': ref_locus.iloc[0]['type1'],\n",
    "                'type2': ref_locus.iloc[0]['type2'],\n",
    "                'name': ref_locus.iloc[0]['name'],\n",
    "                'locus_tag': ref_locus.iloc[0]['locus_tag'],\n",
    "                'product': ref_locus.iloc[0]['product'],\n",
    "                'old_name' : row['name'],\n",
    "                'old_locus_tag' : row['locus_tag'],\n",
    "                'old_product' : row['product']\n",
    "            }, index=[0])], ignore_index=True)\n",
    "\n",
    "# Modify the first column name to prepend '#'\n",
    "# This is to get the '#' as first character in the header\n",
    "new_header = ['#' + improved_bed.columns[0]] + list(improved_bed.columns[1:])\n",
    "\n",
    "# Save the improved bed file\n",
    "improved_bed.to_csv('improved_bed_file_MelihFASTA.bed', sep='\\t', header=new_header, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction</th>\n",
       "      <th>type1</th>\n",
       "      <th>type2</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>old_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>old_product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>410</td>\n",
       "      <td>1750</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gene</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>BSU_00010</td>\n",
       "      <td>chromosomal replication initiator informationa...</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>EO946_RS00002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>410</td>\n",
       "      <td>1750</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gene</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>BSU_00010</td>\n",
       "      <td>chromosomal replication initiator informationa...</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>EO946_RS00003</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>1938</td>\n",
       "      <td>3074</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gene</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>BSU_00020</td>\n",
       "      <td>DNA polymerase III (beta subunit)</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>EO946_RS00004</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>1938</td>\n",
       "      <td>3074</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gene</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>BSU_00020</td>\n",
       "      <td>DNA polymerase III (beta subunit)</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>EO946_RS00005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>3206</td>\n",
       "      <td>3421</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gene</td>\n",
       "      <td>rlbA</td>\n",
       "      <td>BSU_00030</td>\n",
       "      <td>RNA binding protein involved in ribosome matur...</td>\n",
       "      <td>rlbA</td>\n",
       "      <td>EO946_RS00006</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqName  start   end direction type1 type2  name  locus_tag  \\\n",
       "0  NC_000964.3    410  1750         +   CDS  gene  dnaA  BSU_00010   \n",
       "1  NC_000964.3    410  1750         +   CDS  gene  dnaA  BSU_00010   \n",
       "2  NC_000964.3   1938  3074         +   CDS  gene  dnaN  BSU_00020   \n",
       "3  NC_000964.3   1938  3074         +   CDS  gene  dnaN  BSU_00020   \n",
       "4  NC_000964.3   3206  3421         +   CDS  gene  rlbA  BSU_00030   \n",
       "\n",
       "                                             product old_name  old_locus_tag  \\\n",
       "0  chromosomal replication initiator informationa...     dnaA  EO946_RS00002   \n",
       "1  chromosomal replication initiator informationa...     dnaA  EO946_RS00003   \n",
       "2                  DNA polymerase III (beta subunit)     dnaN  EO946_RS00004   \n",
       "3                  DNA polymerase III (beta subunit)     dnaN  EO946_RS00005   \n",
       "4  RNA binding protein involved in ribosome matur...     rlbA  EO946_RS00006   \n",
       "\n",
       "   old_product  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "improved_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tool to convert Snapgene's features list to mock bed file:\n",
    "if False:\n",
    "    \n",
    "    import pandas as pd\n",
    "\n",
    "    def convert_to_bed(input_file, output_file, reference_sequence, header_string=None):\n",
    "        '''\n",
    "        Convert a txt file with gene coordinates to a BED file.\n",
    "        The txt file should have two columns: gene name and coordinates.\n",
    "        The coordinates should be in the format \"start..stop\".\n",
    "        The BED file will have the following columns:\n",
    "        - reference\n",
    "        - start\n",
    "        - stop\n",
    "        - strand\n",
    "        - feature_type\n",
    "\n",
    "        Parameters:\n",
    "        input_file (str): Path to the input txt file.\n",
    "        output_file (str): Path to the output BED file.\n",
    "        reference_sequence (str): Name of the reference sequence.\n",
    "        header_string (str): Optional header string to be added to the BED file.\n",
    "        \n",
    "        Returns:\n",
    "        None\n",
    "        '''\n",
    "        \n",
    "        # Read the txt file\n",
    "        data = pd.read_csv(input_file, sep=r'\\t', header=None, names=['gene', 'coordinates'], engine=\"python\", skipinitialspace=True)\n",
    "\n",
    "        # Split the coordinates into start and stop\n",
    "        data[['start', 'stop']] = data['coordinates'].str.replace(\"..\",\"+\").str.replace(\".\",\"\").str.split('+', expand=True).astype(int)\n",
    "\n",
    "        # Prepare the bed format\n",
    "        bed_data = pd.DataFrame({\n",
    "            'reference': reference_sequence,\n",
    "            'start': data['start'],\n",
    "            'stop': data['stop'],\n",
    "            'strand': '+', # dumbly just put + for all\n",
    "            'feature_type': 'CDS', # dumbly just put CDS for all\n",
    "            'feature': 'gene', # dumbly just put gene for all\n",
    "            'gene_name': data['gene'].str.rstrip(), # remove trailing whitespaces\n",
    "            'gene_id': ['EO946_RS' + str(i).zfill(5) for i in range(1, len(data) + 1)],\n",
    "            'description': \"\" # ['description for ' + gene for gene in data['gene']]\n",
    "        })\n",
    "\n",
    "        # Convert DataFrame to string with tab-separated values\n",
    "        bed_content = bed_data.to_csv(sep='\\t', header=False, index=False, lineterminator=\"\\r\\n\")\n",
    "\n",
    "        # Combine header and BED content\n",
    "        if header_string is None:\n",
    "            full_content = bed_content\n",
    "        else:\n",
    "            full_content = header_string + \"\\r\\n\" + bed_content\n",
    "\n",
    "        # Write to a new BED file\n",
    "        with open(output_file, 'w', newline=\"\\n\") as file:\n",
    "            file.write(full_content)\n",
    "\n",
    "\n",
    "    # Example usage\n",
    "    header_string = \"# this bed file was created by hand from Snapgene's feature export from W23_NCBI_comK_inducible_delta_comK, created by Melih in November 2024\\r\\n\\\n",
    "# Bacillus spizizenii ATCC 6633 = JCM 2499 strain ATCC 6633 chromosome, complete genome. NZ_CP034943\\r\\n\\\n",
    "# seqName\tstart\tende\tdirection\ttype1\ttype2\tName\tlocus_tag\tproduct\"\n",
    "    \n",
    "    convert_to_bed(\"U:/home/philipp/dicts/Features from W23_NCBI_comK_inducible_delta_comK.txt\", \"U:/home/philipp/dicts/Features from W23_NCBI_comK_inducible_delta_comK.bed\", 'NZ_CP034943.1', header_string)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scipy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
